### PR TITLE
fix: fix prefix lookup in basickv and add corresponding test

### PR DESCRIPTION
**Pull request by SYaoJun**
## What's changed and how does it work?

1. add prefixlookup based on pessimistic iterator
2. add test for prefixlookup #133

### CODE REVIEW

File: src/btree/BasicKV.cpp [hunk 1]
```diff
@@ -178,42 +179,31 @@ OpCode BasicKV::Insert(Slice key, Slice val) {
...
+    auto iter = GetIterator();
+    if (iter.SeekToFirstGreaterEqual(prefixKey); !iter.Valid()) {
+      JUMPMU_RETURN OpCode::kNotFound;
+    }
+    bool foundPrefixKey = false;
+    for (; iter.Valid(); iter.Next()) {
+      iter.AssembleKey();
+      auto key = iter.Key();
+      auto value = iter.Val();
+      auto key_string = key.ToString();
```
Comment (zz-jason) [inline src/btree/BasicKV.cpp:193]: better to compare the prefix strings in place to save memory allocation and performance. You can first compare the key size and use `strncmp` to compare bytes content if needed.

File: src/btree/BasicKV.cpp [hunk 2]
```diff
@@ -178,42 +179,31 @@ OpCode BasicKV::Insert(Slice key, Slice val) {
...
-        auto fullKeySize = guardedLeaf->GetFullKeyLen(cur);
-        auto fullKeyBuf = utils::JumpScopedArray<uint8_t>(fullKeySize);
-        guardedLeaf->CopyFullKey(cur, fullKeyBuf->get());
-        guardedLeaf.JumpIfModifiedByOthers();
-
-        callback(Slice(fullKeyBuf->get(), fullKeySize), guardedLeaf->Value(cur));
-        guardedLeaf.JumpIfModifiedByOthers();
-
-        JUMPMU_RETURN OpCode::kOK;
+OpCode BasicKV::PrefixLookup(Slice prefixKey, PrefixLookupCallback callback) {
```
Comment (zz-jason) [inline src/btree/BasicKV.cpp:182]: After going through the old implementation, I think this API can be used as a primitive operation to implement optimistic ScanAsc:

```cpp
actualKey = "";
actualVal = "";

auto callback = [&](key, val) {
  actualKey = key;
  actualVal = val;
}

startKey = someKey;
while (true) {
  auto found = PrefixLookup(startKey, callback) == OpCode::kOK;
  if (found) {
    startKey = actualKey.append(0)
  } else {
    break;
  }
}
```

If so, I think we can implement a helper API like `OptimisticScanAsc(Slice startKey, ScanCallback callback)` for better performance and rename this function to `OptimisticSeekToFirstGreaterEqual`.

Comment (SYaoJun) [reply src/btree/BasicKV.cpp:182]: ok, I got it. I will compose another pull request to resolve this problem.

File: src/btree/BasicKV.cpp (hunk 1 above)
Comment (SYaoJun) [reply src/btree/BasicKV.cpp:193]: I thought about referring to this code on BTreeNode::LowerBound()：
```
if ((key.size() < mPrefixSize) || (bcmp(key.data(), LowerFenceAddr(), mPrefixSize) != 0)) {
      return -1;
    }
```

**Pull request merged**